server/src/main/scala/org/http4s/server/middleware/authentication/BasicAuthentication.scala

```diff
@@ -0,0 +1,42 @@
+package org.http4s
+package server
+package middleware
+package authentication
+
+import org.http4s.headers.Authorization
+
+/**
+ * Provides Basic Authentication from RFC 2617.
+ * @param realm The realm used for authentication purposes.
+ * @param store A partial function mapping (realm, user) to the
+ *              appropriate password.
+ */
+class BasicAuthentication(realm: String, store: AuthenticationStore) extends Authentication {
+  def getChallenge(req: Request): Option[Challenge] = checkAuth(req) match {
+    case OK => None
+    case _ => Some(Challenge("Basic", realm, Nil.toMap))
+  }
+
+  private def checkAuth(req: Request): AuthReply = {
+    req.headers.foldLeft(NoAuthorizationHeader: AuthReply) {
```

Comment:
Could ```headers.get(Authorization)``` and working with the ```Option``` be used here instead of ```foldLeft```? I imagine it would reduce the complexity and allow early escape once the header is found.